tools/souper-check.cpp

```diff
@@ -96,6 +128,74 @@ int SolveInst(const MemoryBufferRef &MB, Solver *S) {
   int Ret = 0;
   int Success = 0, Fail = 0, Error = 0;
   for (auto Rep : Reps) {
+    if (InferNeg) {
+      bool Negative;
+      if (std::error_code EC = S->negative(Rep.BPCs, Rep.PCs, Rep.Mapping.LHS,
+                                           Negative, IC)) {
+        llvm::errs() << EC.message() << '\n';
+      }
+      llvm::outs() << "known negative from souper: " << convertToStr(Negative) << "\n";
+      return 0;
+    }
+
+    if (InferNonNeg) {
+      bool NonNegative;
+      if (std::error_code EC = S->nonNegative(Rep.BPCs, Rep.PCs, Rep.Mapping.LHS,
+                                              NonNegative, IC)) {
+        llvm::errs() << EC.message() << '\n';
+      }
+      llvm::outs() << "known nonNegative from souper: " << convertToStr(NonNegative) << "\n";
+      return 0;
+    }
+
+    if (InferKnownBits) {
+      unsigned W = Rep.Mapping.LHS->Width;
+      KnownBits Known(W);
+      if (std::error_code EC = S->knownBits(Rep.BPCs, Rep.PCs, Rep.Mapping.LHS,
+                                            Known, IC)) {
+        llvm::errs() << EC.message() << '\n';
+      }
+      llvm::outs() << "knownBits from souper: "
+                   << Inst::getKnownBitsString(Known.Zero, Known.One) << "\n";
+      return 0;
+    }
+
+    if (InferPowerTwo) {
+      bool PowTwo;
+      if (std::error_code EC = S->powerTwo(Rep.BPCs, Rep.PCs, Rep.Mapping.LHS,
+                                              PowTwo, IC)) {
+        llvm::errs() << EC.message() << '\n';
+      }
+      llvm::outs() << "known powerOfTwo from souper: " << convertToStr(PowTwo) << "\n";
+      return 0;
+    }
+
+    if (InferNonZero) {
+      bool NonZero;
+      if (std::error_code EC = S->nonZero(Rep.BPCs, Rep.PCs, Rep.Mapping.LHS,
+                                              NonZero, IC)) {
+        llvm::errs() << EC.message() << '\n';
```

Comment:
in all of these functions you must check the error code, not just print it out. the dataflow result is not valid when there's an error, so the code can't just keep going.

Reply:
Fixed.